teps/0046-colocation-of-tasks-and-workspaces.md

```diff
@@ -0,0 +1,123 @@
+
+Before the Affinity Assistant was introduced in [#2630](https://github.com/tektoncd/pipeline/pull/2630) (problem described in [#2586](https://github.com/tektoncd/pipeline/issues/2586)),
+running parallel Tasks concurrently that shared a workspace was difficult. Part of the reason was that
+each Task was run in its own Pod and then needed to use a Persistent Volume to back the Workspace
+and the most commonly available Persistent Volumes has the [access mode](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) `ReadWriteOnce`, so it can only
+be mounted at one node at a time. This made it difficult to run parallel Tasks concurrently since
+they were scheduled to any node and most likely to different nodes (when run in a cluster), and therefore forced to run in a
+**sequence**.
+
+#### With the Affinity Assistant
+
+With the Affinity Assistant, a utility Pod is scheduled to any node and mount the PVC for the workspace.
+Then the Tasks that use the Workspace gets Pod-Affinity to the utility Pod so that these Pods are
+scheduled to the **same node**. This is a technique to co-schedule Tasks using a workspace to the same
+node were the workspace is already mounted and since they are on the same node, they can concurrently
+access the Persisent Volume even if it has access mode `ReadWriteOnce`.
+
+This solution helped to run the parallel Tasks concurrently, but the solution has shortcomings. One 
+problem is that the Pods must always be scheduled to the same node as the utility Pod, even if that
+node is out of resources, which means that the new Pod will stay in state Pending, potentially for long time.
+This solution also can not use auto-scaling in the cluster, since the new Pod is restricted to be 
+scheduled to the same node as the utility Pod even if the cluster could auto-scale with more nodes.
+
+With this solution, Persistent Volumes still has to be used for the Workspace, since the Tasks
+run in its own Pods. Using Persistent Volume has other drawbacks than described above, they. e.g. are not available 
+on all clusters, they need to be cleaned up, they has an economic cost and also a performance cost.
+
+### Proposal for Colocation of Tasks and Workspaces
+
+To avoid the problems with the above described solutions, this enhancement proposal is about introducing
+a solution that uses a Pod-internal workspace and colocate the Tasks that use the workspace within
+the same scheduling unit (in Kubernetes, a Pod).
+
+A new constraint with this solution is that only a single ServiceAccount can be used for the Tasks within
+the same scheduling unit (Pod).
+
+### Goals
+
+- Make it possible to use a Pod-internal workspace to share data in a PipelineRun
+    - The Tasks that use the workspace is scheduled to run within the same scheduling unit (Pod)
+    - That Pipeline-features in use today is still usable, e.g. concurrency and `When`-expressions
+- No changes in the Pipeline or Task API for authors
```

Comment:
@jlpettersson I forgot to ask you in the API working group: does this mean that you're proposing that this is something that is configured only at runtime (i.e. in the PipelineRun only)? or would you imagine the pipeline author expressing this also/instead?